src/client/connection.ts

```diff
@@ -40,23 +40,36 @@ export const eventEmitter = new EventEmitter();
 /**
  * Promise to the channel object that allows us to connect to the client
  */
-export let channelPromise: Promise<ChannelClient>;
-// Currently a runtime bug when provider connects to itself. Ideally the provider would never import a file
-// that includes this, but for now it is easier to put a guard in place.
-if (fin.Window.me.uuid !== SERVICE_IDENTITY.uuid || fin.Window.me.name !== SERVICE_IDENTITY.name) {
-    channelPromise = typeof fin === 'undefined' ?
-        Promise.reject(new Error('fin is not defined. The openfin-fdc3 module is only intended for use in an OpenFin application.')) :
-        fin.InterApplicationBus.Channel.connect(SERVICE_CHANNEL, {payload: {version: PACKAGE_VERSION}}).then((channel: ChannelClient) => {
-            // Register service listeners
-            channel.register('WARN', (payload: any) => console.warn(payload));  // tslint:disable-line:no-any
-            channel.register('event', (event: FDC3Event) => {
-                eventEmitter.emit(event.type, event);
-            });
-            // Any unregistered action will simply return false
-            channel.setDefaultAction(() => false);
+let channelPromise: Promise<ChannelClient>|null = null;
 
-            return channel;
-        });
+if (typeof fin !== 'undefined') {
+    getServicePromise();
+}
+
+export function getServicePromise(): Promise<ChannelClient> {
+    if (!channelPromise) {
+        // Currently a runtime bug when provider connects to itself. Ideally the provider would never import a file
+        // that includes this, but for now it is easier to put a guard in place.
+        if (fin.Window.me.uuid !== SERVICE_IDENTITY.uuid || fin.Window.me.name !== SERVICE_IDENTITY.name) {
+            channelPromise = typeof fin === 'undefined' ?
```

Comment:
I spotted this a while ago but now that it's being rewritten could be a good time to fix.

This `undefined` check after accessing `fin.Window....` doesn't make sense. If we want to throw the custom error it should happen right before (and maybe make redundant the extra `if` on line 45?)

Reply:
Cool, fixed. We can't get rid of the earlier `'undefined'' check, as we do want to do that early getServicePromise() call, but _only_ within the OpenFin environment